wellcomeml/ml/transformers_tokenizer.py

```diff
@@ -0,0 +1,95 @@
+"""
+Implements Tokenizer that abstracts common tokenisation
+strategies used in Transformers
+"""
+from tokenizers.models import WordPiece, BPE
+from tokenizers.normalizers import Lowercase, Sequence
+from tokenizers.trainers import BpeTrainer, WordPieceTrainer
+from tokenizers.pre_tokenizers import ByteLevel, Whitespace
+from tokenizers import Tokenizer
+
+
+# TODO
+# - generalise to two sentences
+# - prepare input for transformers
+# - pad and truncate to max length
+
+class TransformersTokenizer:
+    def __init__(self, lowercase=True,
+                 pre_tokenizer="whitespace", model="wordpiece",
+                 vocab_size=30_000, unk_token="[UNK]"):
+        self.lowercase = lowercase
+        self.pre_tokenizer = pre_tokenizer
+        self.model = model
+        self.vocab_size = vocab_size
+        self.unk_token = unk_token
+
+    def __getstate__(self):
+        attributes = self.__dict__.copy()
+        del attributes['trainer']
+        return attributes
+
+    def _init_tokenizer(self):
+        if self.model == "wordpiece":
+            model = WordPiece(unk_token=self.unk_token)
+            self.trainer = WordPieceTrainer(
+                vocab_size=self.vocab_size,
+                special_tokens=[self.unk_token])
+        elif self.model == "bpe":
+            model = BPE(unk_token=self.unk_token)
+            self.trainer = BpeTrainer(
+                vocab_size=self.vocab_size,
+                special_tokens=[self.unk_token])
+        else:
+            raise NotImplementedError
+
+        normalizers = []
+        if self.lowercase:
+            normalizers.append(Lowercase())
+        normalizers = Sequence(normalizers)
+
+        if self.pre_tokenizer == "bytelevel":
+            pre_tokenizer = ByteLevel()
+        elif self.pre_tokenizer == "whitespace":
+            pre_tokenizer = Whitespace()
+        else:
+            raise NotImplementedError
+
+        self.tokenizer = Tokenizer(model)
+        self.tokenizer.normalizer = normalizers
+        self.tokenizer.pre_tokenizer = pre_tokenizer
+
+    @property
+    def vocab(self):
+        return self.tokenizer.get_vocab()
+
+    def fit(self, texts):
+        self._init_tokenizer()
+        self.tokenizer.train_from_iterator(texts, trainer=self.trainer)
```

Comment:
the tests fail because of this ('tokenizers.Tokenizer' object has no attribute 'train_from_iterator').
Seems like you might need to pin tokenizers 0.10.0rc1, see:
https://github.com/huggingface/tokenizers/issues/548

Reply:
i know i know, thanks for checking. problem is that 0.10 is incompatible with the transformers version